common/djangoapps/student/models.py

```diff
@@ -3116,6 +3122,135 @@ class AccountRecoveryConfiguration(ConfigurationModel):
+    2. If a learner misses enough days to count as a break, the streak resets back to 0.
+    3. The streak is measured against the learner's configured timezone
+    4. We keep track of the total length of the streak, so there is a possibility in the future
+    to add multiple celebrations for longer streaks.
+    5. We keep track of the longest_ever_streak field for potential future use for badging purposes.
+
+    .. no_pii:
+    """
+    user = models.OneToOneField(User, models.CASCADE, related_name='celebration')
+    first_day_of_streak = models.DateField(default=None, null=True, blank=True)
+    last_day_of_streak = models.DateField(default=None, null=True, blank=True)
+    last_streak_celebration = models.DateField(default=None, null=True, blank=True)
+    longest_ever_streak = models.IntegerField(default=0)
+
+    def __str__(self):
+        return (
+            "[CourseEnrollmentCelebration] user: {}; first_day_of_streak {};"
+            " last_day_of_streak {}; last_streak_celebration {}; longest_ever_streak {};"
+        ).format(self.user.username, self.first_day_of_streak,
+                 self.last_day_of_streak, self.last_streak_celebration, self.longest_ever_streak)
+
+    @classmethod
+    def _get_now(cls):
+        """ Retrieve the value for the current datetime in the user's timezone """
+        # importing here to avoid a circular import
+        from lms.djangoapps.courseware.context_processor import user_timezone_locale_prefs
+        user_timezone_locale = user_timezone_locale_prefs(crum.get_current_request())
+        user_timezone = timezone(user_timezone_locale['user_timezone'] or str(UTC))
+        return user_timezone.localize(datetime.now())
+
+    @classmethod
+    def _calculate_streak_updates(cls, celebration, now):
+        """ Calculate the updates that should me applied to the streak fields of the provided celebration """
+        first_day_of_streak = celebration.first_day_of_streak
+        last_day_of_streak = celebration.last_day_of_streak
+        last_streak_celebration = celebration.last_streak_celebration
+        today = now.date()
+        yesterday = today - timedelta(days=1)
+        should_celebrate = False
+
+        first_ever_streak = None is last_streak_celebration is first_day_of_streak is last_day_of_streak
+        in_active_streak = last_day_of_streak and first_day_of_streak
+        broken_streak = last_day_of_streak not in (yesterday, today)
+        can_start_new_streak = last_day_of_streak and last_day_of_streak + timedelta(days=STREAK_BREAK_LENGTH) < today
+
+        if first_ever_streak:
+            # Start new streak
+            first_day_of_streak = today
+            last_day_of_streak = today
+        elif broken_streak:
+            if can_start_new_streak:
+                # If we were on a break and the break is over start new streak
+                first_day_of_streak = today
+                last_day_of_streak = today
+        elif in_active_streak:
```

Comment:
This could be an `else` right? There's not a fourth state this celebration could be in, is there?

Reply:
Maybe I don't know what inside a break means. Because it looks like `inside_active_streak` simply requires a non-`None` `last_day_of_streak` and `first_day_of_streak` - which is basically the opposite of `first_ever_streak` (unless one of those day variables can be set and the other not...?)

To be able to be "in the middle of the break", would we need to use `break_length` inside the definition of `inside_active_streak`?

Reply:
ah i got confused and incorrectly removed broken_streak in response to this comment
https://github.com/edx/edx-platform/pull/26409#discussion_r577022136
i put it back and renamed it missed_day_of_streak which might be clearer
so in the middle of a break would be at the spot where
missed_day_of_streak is True
but 
can_start_new_streak is False